File: Source/Client/Modules/Network/NetworkConfig.vb
Imports Mirage.Sharp.Asfw.Network
Imports Core

Friend Module NetworkConfig
    Friend WithEvents Socket As NetworkClient

    Friend Sub InitNetwork()
        Socket = New NetworkClient(ServerPackets.COUNT, 8192)
        Connect()
        PacketRouter()
    End Sub

    Friend Sub Connect()
        Socket?.Connect(Type.Setting.Ip, Type.Setting.Port)
    End Sub

    Friend Sub DestroyNetwork()
        ' Calling a disconnect is not necessary when using destroy network as
        ' Dispose already calls it and cleans up the memory internally.
        Socket?.Dispose()
    End Sub

#Region " Events "

    Private Sub Socket_ConnectionSuccess() Handles Socket.ConnectionSuccess

    End Sub

    Private Sub Socket_ConnectionFailed() Handles Socket.ConnectionFailed
        DestroyNetwork()
        InitNetwork()
    End Sub

    Private Sub Socket_ConnectionLost() Handles Socket.ConnectionLost
        
    End Sub

    Private Sub Socket_CrashReport(err As String) Handles Socket.CrashReport
        LogoutGame()
        DialogueAlert(DialogueMsg.Crash)

        Dim currentDateTime As DateTime = DateTime.Now
        Dim timestampForFileName As String = currentDateTime.ToString("yyyyMMdd_HHmmss")
        Dim logFileName As String = $"{timestampForFileName}.txt"

        Addlog(err, logFileName)
    End Sub

#If DEBUG Then

    Private Sub Socket_TrafficReceived(size As Integer, ByRef data() As Byte) Handles Socket.TrafficReceived
        Console.WriteLine("Traffic Received : [Size: " & size & "]")
        Dim tmpData = data
#Disable Warning BC42024 ' Unused local variable
        Dim breakPointDummy As Integer
#Enable Warning BC42024 ' Unused local variable
        'Put breakline on BreakPointDummy to look at what is contained in data at runtime in the VS logger.
    End Sub

    Private Sub Socket_PacketReceived(size As Integer, header As Integer, ByRef data() As Byte) Handles Socket.PacketReceived
        Console.WriteLine("Packet Received : [Size: " & size & "| Packet: " & CType(header, ServerPackets).ToString() & "]")
        Dim tmpData = data
#Disable Warning BC42024 ' Unused local variable
        Dim breakPointDummy As Integer
#Enable Warning BC42024 ' Unused local variable
        'Put breakline on BreakPointDummy to look at what is contained in data at runtime in the VS logger.
    End Sub

#End If

#End Region

End Module

File: Source/Client/Game/Network/NetworkConfig.cs
using System;
using System.Runtime.CompilerServices;
using Core;
using Mirage.Sharp.Asfw.Network;

namespace Client
{

    public class NetworkConfig
    {
        private static NetworkClient _Socket;

        public static NetworkClient Socket
        {
            [MethodImpl(MethodImplOptions.Synchronized)]
            get
            {
                return _Socket;
            }

            [MethodImpl(MethodImplOptions.Synchronized)]
            set
            {
                if (_Socket != null)
                {
                    _Socket.ConnectionSuccess -= Socket_ConnectionSuccess;
                    _Socket.ConnectionFailed -= Socket_ConnectionFailed;
                    _Socket.ConnectionLost -= Socket_ConnectionLost;
                    _Socket.CrashReport -= Socket_CrashReport;
                    _Socket.TrafficReceived -= Socket_TrafficReceived;
                    _Socket.PacketReceived -= Socket_PacketReceived;
                }

                _Socket = value;
                if (_Socket != null)
                {
                    _Socket.ConnectionSuccess += Socket_ConnectionSuccess;
                    _Socket.ConnectionFailed += Socket_ConnectionFailed;
                    _Socket.ConnectionLost += Socket_ConnectionLost;
                    _Socket.CrashReport += Socket_CrashReport;
                    _Socket.TrafficReceived += Socket_TrafficReceived;
                    _Socket.PacketReceived += Socket_PacketReceived;
                }
            }
        }

        public static void InitNetwork()
        {
            try
            {
                // Initialize the network client with packet count and buffer size.
                Socket = new NetworkClient((int)Packets.ServerPackets.COUNT, 8192);

                // Start the connection attempt.
                Socket.ConnectionSuccess += OnConnectionSuccess;

                Socket.Connect(Settings.Instance.IP, Settings.Instance.Port); // Adjust IP and port as needed
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Network initialization failed: {ex.Message}");
            }
        }

        private static void OnConnectionSuccess()
        {
            Console.WriteLine("Connection established. Starting packet router...");
            NetworkReceive.PacketRouter();
        }

        public static void DestroyNetwork()
        {
            // Calling a disconnect is not necessary when using destroy network as
            // Dispose already calls it and cleans up the memory internally.
            Socket?.Dispose();
        }

        #region  Events 

        private static void Socket_ConnectionSuccess()
        {
            Console.WriteLine("Connection success.");
        }

        private static void Socket_ConnectionFailed()
        {
            Console.WriteLine("Failed to connect to the server. Retrying...");
            InitNetwork();
        }

        private static void Socket_ConnectionLost()
        {
            Console.WriteLine("Connection lost.");
        }

        private static void Socket_CrashReport(string err)
        {
            GameLogic.LogoutGame();
            GameLogic.DialogueAlert((byte)Core.Enum.DialogueMsg.Crash);

            var currentDateTime = DateTime.Now;
            string timestampForFileName = currentDateTime.ToString("yyyyMMdd_HHmmss");
            string logFileName = $"{timestampForFileName}.txt";

            Core.Log.Add(err, logFileName);
        }

        private static void Socket_TrafficReceived(int size, ref byte[] data)
        {
            Console.WriteLine("Traffic Received : [Size: " + size + "]");
            byte[] tmpData = data;
            // Put breakline on tmpData to look at what is contained in data at runtime in the VS logger.
        }

        private static void Socket_PacketReceived(int size, int header, ref byte[] data)
        {
            Console.WriteLine("Packet Received : [Size: " + size + "| Packet: " + ((Packets.ServerPackets)header).ToString() + "]");
            byte[] tmpData = data;
            // Put breakline on tmpData to look at what is contained in data at runtime in the VS logger.
        }
        #endregion

    }
}